File: generate_icon.py
#!/usr/bin/env python3
"""Generate the Weather Wise app icon: cloud morphing into brain."""

from PIL import Image, ImageDraw
import math

SIZE = 1024
img = Image.new("RGBA", (SIZE, SIZE))
draw = ImageDraw.Draw(img)
w = h = SIZE


def lerp_color(c1, c2, t):
    return tuple(int(a + (b - a) * t) for a, b in zip(c1, c2))


# MARK: - Background gradient (top-left to bottom-right)
for y in range(h):
    for x in range(w):
        t = (x / w + y / h) / 2.0
        if t < 0.5:
            c = lerp_color((38, 102, 217), (20, 51, 140), t * 2)
        else:
            c = lerp_color((20, 51, 140), (13, 31, 89), (t - 0.5) * 2)
        img.putpixel((x, y), (*c, 255))

# Radial glow
cx, cy = w * 0.5, h * 0.48
max_r = w * 0.4
for y in range(int(h * 0.15), int(h * 0.82)):
    for x in range(int(w * 0.15), int(w * 0.85)):
        dx, dy = x - cx, y - cy
        r = math.sqrt(dx * dx + dy * dy)
        if r < max_r:
            t = r / max_r
            alpha = int(30 * (1.0 - t * t))
            if alpha > 0:
                px = img.getpixel((x, y))
                blended = tuple(min(255, c + alpha) for c in px[:3])
                img.putpixel((x, y), (*blended, 255))


# MARK: - Draw cloud-brain using polygon approximation
def bezier(p0, p1, p2, p3, steps=30):
    """Cubic bezier curve points."""
    pts = []
    for i in range(steps + 1):
        t = i / steps
        u = 1 - t
        x = u**3 * p0[0] + 3 * u**2 * t * p1[0] + 3 * u * t**2 * p2[0] + t**3 * p3[0]
        y = u**3 * p0[1] + 3 * u**2 * t * p1[1] + 3 * u * t**2 * p2[1] + t**3 * p3[1]
        pts.append((x, y))
    return pts


def build_cloud_brain():
    baseY = h * 0.68
    baseLeft = w * 0.18
    baseRight = w * 0.82
    pts = [(baseLeft, baseY)]

    # Left cloud puffs
    pts += bezier((baseLeft, baseY), (w*0.10, baseY), (w*0.08, h*0.58), (w*0.14, h*0.52))
    pts += bezier((w*0.14, h*0.52), (w*0.10, h*0.44), (w*0.10, h*0.36), (w*0.18, h*0.34))
    pts += bezier((w*0.18, h*0.34), (w*0.22, h*0.28), (w*0.26, h*0.24), (w*0.32, h*0.24))

    # Top transition
    pts += bezier((w*0.32, h*0.24), (w*0.38, h*0.22), (w*0.43, h*0.20), (w*0.48, h*0.22))
    # Brain fissure dip
    pts += bezier((w*0.48, h*0.22), (w*0.50, h*0.25), (w*0.51, h*0.25), (w*0.52, h*0.23))

    # Right brain lobes
    pts += bezier((w*0.52, h*0.23), (w*0.57, h*0.19), (w*0.63, h*0.20), (w*0.68, h*0.24))
    pts += bezier((w*0.68, h*0.24), (w*0.74, h*0.24), (w*0.80, h*0.26), (w*0.80, h*0.32))
    pts += bezier((w*0.80, h*0.32), (w*0.83, h*0.35), (w*0.84, h*0.38), (w*0.82, h*0.40))
    pts += bezier((w*0.82, h*0.40), (w*0.86, h*0.44), (w*0.88, h*0.48), (w*0.86, h*0.52))
    pts += bezier((w*0.86, h*0.52), (w*0.88, h*0.60), (w*0.88, baseY), (baseRight, baseY))

    pts.append((baseLeft, baseY))
    return pts


shape = build_cloud_brain()
shape_tuples = [tuple(map(int, p)) for p in shape]

# Shadow
shadow_pts = [(int(p[0] + 4), int(p[1] + 6)) for p in shape]
shadow_img = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
shadow_draw = ImageDraw.Draw(shadow_img)
shadow_draw.polygon(shadow_pts, fill=(0, 0, 0, 60))
img = Image.alpha_composite(img, shadow_img)

# Shape fill with left-to-right gradient (cloud blue -> brain pink)
# Create a mask from the shape
mask = Image.new("L", (SIZE, SIZE), 0)
mask_draw = ImageDraw.Draw(mask)
mask_draw.polygon(shape_tuples, fill=255)

fill_layer = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
for y in range(h):
    for x in range(w):
        if mask.getpixel((x, y)) > 0:
            t = (x - w * 0.15) / (w * 0.7)
            t = max(0.0, min(1.0, t))
            if t < 0.5:
                c = lerp_color((217, 235, 255), (242, 224, 242), t * 2)
            else:
                c = lerp_color((242, 224, 242), (235, 209, 230), (t - 0.5) * 2)
            # Top highlight
            shape_top = h * 0.20
            shape_bot = h * 0.68
            yt = (y - shape_top) / (shape_bot - shape_top)
            if yt < 0.4:
                highlight = int(120 * (1.0 - yt / 0.4))
                c = tuple(min(255, v + highlight) for v in c)
            fill_layer.putpixel((x, y), (*c, 255))

img = Image.alpha_composite(img, fill_layer)

# Shape outline
outline_img = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
outline_draw = ImageDraw.Draw(outline_img)
outline_draw.polygon(shape_tuples, outline=(255, 255, 255, 100))
# Draw it a couple times for thickness
for offset in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
    shifted = [(int(p[0] + offset[0]), int(p[1] + offset[1])) for p in shape]
    outline_draw.polygon(shifted, outline=(255, 255, 255, 60))
img = Image.alpha_composite(img, outline_img)


# MARK: - Brain sulci (fold lines)
def draw_bezier_line(draw_obj, p0, p1, p2, p3, color, width=8):
    pts = bezier(p0, p1, p2, p3, steps=40)
    for i in range(len(pts) - 1):
        draw_obj.line([pts[i], pts[i+1]], fill=color, width=width)


sulci_img = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
sulci_draw = ImageDraw.Draw(sulci_img)
sulci_color = (178, 153, 184, 128)
lw = 8

# Central sulcus
draw_bezier_line(sulci_draw,
    (w*0.50, h*0.25), (w*0.51, h*0.38), (w*0.50, h*0.52), (w*0.52, h*0.62),
    sulci_color, lw)
# Upper fold
draw_bezier_line(sulci_draw,
    (w*0.54, h*0.32), (w*0.62, h*0.28), (w*0.70, h*0.30), (w*0.76, h*0.33),
    sulci_color, lw)
# Middle fold
draw_bezier_line(sulci_draw,
    (w*0.53, h*0.43), (w*0.62, h*0.40), (w*0.74, h*0.40), (w*0.80, h*0.44),
    sulci_color, lw)
# Lower fold
draw_bezier_line(sulci_draw,
    (w*0.53, h*0.54), (w*0.63, h*0.51), (w*0.75, h*0.51), (w*0.82, h*0.55),
    sulci_color, lw)

# Mask sulci to shape
sulci_masked = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
sulci_masked.paste(sulci_img, mask=mask)
img = Image.alpha_composite(img, sulci_masked)


# MARK: - Lightning bolt
bolt_pts = [
    (w*0.50, h*0.60),
    (w*0.46, h*0.70),
    (w*0.49, h*0.70),
    (w*0.46, h*0.80),
    (w*0.54, h*0.72),
    (w*0.51, h*0.72),
    (w*0.55, h*0.60),
]
bolt_tuples = [tuple(map(int, p)) for p in bolt_pts]

bolt_img = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
bolt_draw = ImageDraw.Draw(bolt_img)

# Bolt fill gradient (yellow to orange, top to bottom)
bolt_mask = Image.new("L", (SIZE, SIZE), 0)
bolt_mask_draw = ImageDraw.Draw(bolt_mask)
bolt_mask_draw.polygon(bolt_tuples, fill=255)

for y in range(int(h * 0.58), int(h * 0.82)):
    for x in range(int(w * 0.44), int(w * 0.57)):
        if bolt_mask.getpixel((x, y)) > 0:
            t = (y - h * 0.58) / (h * 0.22)
            t = max(0.0, min(1.0, t))
            c = lerp_color((255, 217, 51), (255, 166, 26), t)
            bolt_img.putpixel((x, y), (*c, 255))

img = Image.alpha_composite(img, bolt_img)

# Bolt outline
bolt_outline = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
bolt_outline_draw = ImageDraw.Draw(bolt_outline)
bolt_outline_draw.polygon(bolt_tuples, outline=(255, 153, 0, 153))
for offset in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
    shifted = [(int(p[0] + offset[0]), int(p[1] + offset[1])) for p in bolt_pts]
    bolt_outline_draw.polygon(shifted, outline=(255, 153, 0, 100))
img = Image.alpha_composite(img, bolt_outline)


# Save
output = "WeatherWise/Assets.xcassets/AppIcon.appiconset/AppIcon.png"
img.save(output, "PNG")
print(f"App icon saved to {output} ({SIZE}x{SIZE})")

File: generate_icon.swift
#!/usr/bin/env swift

import Foundation
import CoreGraphics
import ImageIO
import UniformTypeIdentifiers

let size = 1024
let w = CGFloat(size)
let h = CGFloat(size)

let colorSpace = CGColorSpaceCreateDeviceRGB()
guard let ctx = CGContext(
    data: nil,
    width: size,
    height: size,
    bitsPerComponent: 8,
    bytesPerRow: 0,
    space: colorSpace,
    bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
) else {
    print("Failed to create context")
    exit(1)
}

// Flip coordinate system (CG is bottom-left origin)
ctx.translateBy(x: 0, y: h)
ctx.scaleBy(x: 1, y: -1)

// MARK: - Background gradient
let bgColors: [CGFloat] = [
    0.15, 0.40, 0.85, 1.0,
    0.08, 0.20, 0.55, 1.0,
    0.05, 0.12, 0.35, 1.0
]
let bgLocations: [CGFloat] = [0.0, 0.5, 1.0]
if let gradient = CGGradient(colorSpace: colorSpace, colorComponents: bgColors, locations: bgLocations, count: 3) {
    ctx.drawLinearGradient(gradient, start: CGPoint(x: 0, y: 0), end: CGPoint(x: w, y: h), options: [])
}

// MARK: - Radial glow
let glowColors: [CGFloat] = [
    1.0, 1.0, 1.0, 0.12,
    1.0, 1.0, 1.0, 0.03,
    1.0, 1.0, 1.0, 0.0
]
let glowLocations: [CGFloat] = [0.0, 0.5, 1.0]
if let glow = CGGradient(colorSpace: colorSpace, colorComponents: glowColors, locations: glowLocations, count: 3) {
    ctx.drawRadialGradient(glow, startCenter: CGPoint(x: w * 0.5, y: h * 0.48), startRadius: 0, endCenter: CGPoint(x: w * 0.5, y: h * 0.48), endRadius: w * 0.4, options: [])
}

// MARK: - Cloud-brain shape
func cloudBrainPath() -> CGMutablePath {
    let path = CGMutablePath()
    let baseY = h * 0.68
    let baseLeft = w * 0.18
    let baseRight = w * 0.82

    path.move(to: CGPoint(x: baseLeft, y: baseY))

    // Left cloud puffs
    path.addCurve(to: CGPoint(x: w * 0.14, y: h * 0.52),
                  control1: CGPoint(x: w * 0.10, y: baseY),
                  control2: CGPoint(x: w * 0.08, y: h * 0.58))
    path.addCurve(to: CGPoint(x: w * 0.18, y: h * 0.34),
                  control1: CGPoint(x: w * 0.10, y: h * 0.44),
                  control2: CGPoint(x: w * 0.10, y: h * 0.36))
    path.addCurve(to: CGPoint(x: w * 0.32, y: h * 0.24),
                  control1: CGPoint(x: w * 0.22, y: h * 0.28),
                  control2: CGPoint(x: w * 0.26, y: h * 0.24))

    // Top transition
    path.addCurve(to: CGPoint(x: w * 0.48, y: h * 0.22),
                  control1: CGPoint(x: w * 0.38, y: h * 0.22),
                  control2: CGPoint(x: w * 0.43, y: h * 0.20))
    // Brain fissure dip
    path.addCurve(to: CGPoint(x: w * 0.52, y: h * 0.23),
                  control1: CGPoint(x: w * 0.50, y: h * 0.25),
                  control2: CGPoint(x: w * 0.51, y: h * 0.25))

    // Right brain lobes
    path.addCurve(to: CGPoint(x: w * 0.68, y: h * 0.24),
                  control1: CGPoint(x: w * 0.57, y: h * 0.19),
                  control2: CGPoint(x: w * 0.63, y: h * 0.20))
    path.addCurve(to: CGPoint(x: w * 0.80, y: h * 0.32),
                  control1: CGPoint(x: w * 0.74, y: h * 0.24),
                  control2: CGPoint(x: w * 0.80, y: h * 0.26))
    path.addCurve(to: CGPoint(x: w * 0.82, y: h * 0.40),
                  control1: CGPoint(x: w * 0.83, y: h * 0.35),
                  control2: CGPoint(x: w * 0.84, y: h * 0.38))
    path.addCurve(to: CGPoint(x: w * 0.86, y: h * 0.52),
                  control1: CGPoint(x: w * 0.86, y: h * 0.44),
                  control2: CGPoint(x: w * 0.88, y: h * 0.48))
    path.addCurve(to: CGPoint(x: baseRight, y: baseY),
                  control1: CGPoint(x: w * 0.88, y: h * 0.60),
                  control2: CGPoint(x: w * 0.88, y: baseY))

    path.addLine(to: CGPoint(x: baseLeft, y: baseY))
    path.closeSubpath()
    return path
}

let shape = cloudBrainPath()

// Shadow
ctx.saveGState()
ctx.translateBy(x: 3, y: 5)
ctx.setFillColor(CGColor(red: 0, green: 0, blue: 0, alpha: 0.25))
ctx.addPath(shape)
ctx.fillPath()
ctx.restoreGState()

// Main fill gradient (cloud-blue to brain-pink)
ctx.saveGState()
ctx.addPath(shape)
ctx.clip()
let fillColors: [CGFloat] = [
    0.85, 0.92, 1.0, 1.0,
    0.95, 0.88, 0.95, 1.0,
    0.92, 0.82, 0.90, 1.0
]
let fillLocations: [CGFloat] = [0.0, 0.5, 1.0]
if let fillGrad = CGGradient(colorSpace: colorSpace, colorComponents: fillColors, locations: fillLocations, count: 3) {
    ctx.drawLinearGradient(fillGrad, start: CGPoint(x: w * 0.15, y: h * 0.5), end: CGPoint(x: w * 0.85, y: h * 0.5), options: [])
}
ctx.restoreGState()

// Highlight on top
ctx.saveGState()
ctx.addPath(shape)
ctx.clip()
let hlColors: [CGFloat] = [
    1.0, 1.0, 1.0, 0.5,
    1.0, 1.0, 1.0, 0.0
]
let hlLocations: [CGFloat] = [0.0, 1.0]
if let hlGrad = CGGradient(colorSpace: colorSpace, colorComponents: hlColors, locations: hlLocations, count: 2) {
    ctx.drawLinearGradient(hlGrad, start: CGPoint(x: w * 0.5, y: h * 0.2), end: CGPoint(x: w * 0.5, y: h * 0.55), options: [])
}
ctx.restoreGState()

// Outline
ctx.setStrokeColor(CGColor(red: 1, green: 1, blue: 1, alpha: 0.4))
ctx.setLineWidth(w * 0.004)
ctx.addPath(shape)
ctx.strokePath()

// MARK: - Brain sulci
ctx.setStrokeColor(CGColor(red: 0.7, green: 0.6, blue: 0.72, alpha: 0.5))
ctx.setLineWidth(w * 0.008)
ctx.setLineCap(.round)

// Central sulcus
let sulci = CGMutablePath()
sulci.move(to: CGPoint(x: w * 0.50, y: h * 0.25))
sulci.addCurve(to: CGPoint(x: w * 0.52, y: h * 0.62),
               control1: CGPoint(x: w * 0.51, y: h * 0.38),
               control2: CGPoint(x: w * 0.50, y: h * 0.52))

// Upper fold
sulci.move(to: CGPoint(x: w * 0.54, y: h * 0.32))
sulci.addCurve(to: CGPoint(x: w * 0.76, y: h * 0.33),
               control1: CGPoint(x: w * 0.62, y: h * 0.28),
               control2: CGPoint(x: w * 0.70, y: h * 0.30))

// Middle fold
sulci.move(to: CGPoint(x: w * 0.53, y: h * 0.43))
sulci.addCurve(to: CGPoint(x: w * 0.80, y: h * 0.44),
               control1: CGPoint(x: w * 0.62, y: h * 0.40),
               control2: CGPoint(x: w * 0.74, y: h * 0.40))

// Lower fold
sulci.move(to: CGPoint(x: w * 0.53, y: h * 0.54))
sulci.addCurve(to: CGPoint(x: w * 0.82, y: h * 0.55),
               control1: CGPoint(x: w * 0.63, y: h * 0.51),
               control2: CGPoint(x: w * 0.75, y: h * 0.51))

ctx.addPath(sulci)
ctx.strokePath()

// MARK: - Lightning bolt
let bolt = CGMutablePath()
bolt.move(to: CGPoint(x: w * 0.50, y: h * 0.60))
bolt.addLine(to: CGPoint(x: w * 0.46, y: h * 0.70))
bolt.addLine(to: CGPoint(x: w * 0.49, y: h * 0.70))
bolt.addLine(to: CGPoint(x: w * 0.46, y: h * 0.80))
bolt.addLine(to: CGPoint(x: w * 0.54, y: h * 0.72))
bolt.addLine(to: CGPoint(x: w * 0.51, y: h * 0.72))
bolt.addLine(to: CGPoint(x: w * 0.55, y: h * 0.60))
bolt.closeSubpath()

// Bolt fill gradient
ctx.saveGState()
ctx.addPath(bolt)
ctx.clip()
let boltColors: [CGFloat] = [
    1.0, 0.85, 0.2, 1.0,
    1.0, 0.65, 0.1, 1.0
]
let boltLocations: [CGFloat] = [0.0, 1.0]
if let boltGrad = CGGradient(colorSpace: colorSpace, colorComponents: boltColors, locations: boltLocations, count: 2) {
    ctx.drawLinearGradient(boltGrad, start: CGPoint(x: w * 0.47, y: h * 0.58), end: CGPoint(x: w * 0.53, y: h * 0.78), options: [])
}
ctx.restoreGState()

// Bolt outline
ctx.setStrokeColor(CGColor(red: 1.0, green: 0.6, blue: 0.0, alpha: 0.6))
ctx.setLineWidth(w * 0.003)
ctx.addPath(bolt)
ctx.strokePath()

// MARK: - Save PNG
guard let image = ctx.makeImage() else {
    print("Failed to create image")
    exit(1)
}

let outputPath = "WeatherWise/Assets.xcassets/AppIcon.appiconset/AppIcon.png"
let url = URL(fileURLWithPath: outputPath)

guard let dest = CGImageDestinationCreateWithURL(url as CFURL, "public.png" as CFString, 1, nil) else {
    print("Failed to create image destination")
    exit(1)
}
CGImageDestinationAddImage(dest, image, nil)
if CGImageDestinationFinalize(dest) {
    print("App icon saved to \(outputPath)")
} else {
    print("Failed to write PNG")
    exit(1)
}
